plugins/http_plugin/common.hpp

```diff
@@ -0,0 +1,276 @@
+         * @return in flight size of o
+         */
+        template<typename T>
+        static size_t in_flight_sizeof( const std::optional<T>& o ) {
+            if( o ) {
+            return in_flight_sizeof( *o );
+            }
+            return 0;
+        }
+
+        /**
+         * Helper method to calculate the "in flight" size of a string
+         * @param s - the string
+         * @return in flight size of s
+         */
+        static size_t in_flight_sizeof( const string& s ) {
+            return s.size();
+        }
+    }
+
+    // key -> priority, url_handler
+    typedef map<string,detail::internal_url_handler> url_handlers_type;
+
+    struct http_plugin_state {
+        string                         access_control_allow_origin;
+        string                         access_control_allow_headers;
+        string                         access_control_max_age;
+        bool                           access_control_allow_credentials = false;
+        size_t                         max_body_size{1024*1024};
+
+        std::atomic<size_t>                            bytes_in_flight{0};
+        std::atomic<int32_t>                           requests_in_flight{0};
+        size_t                                         max_bytes_in_flight = 0;
+        int32_t                                        max_requests_in_flight = -1;
+        fc::microseconds                               max_response_time{30*1000};
+
+        bool                     validate_host = true;
+        set<string>              valid_hosts;
+
+        bool verbose_http_errors = false;
+
+        url_handlers_type  url_handlers;
+        bool                     keep_alive = false;
+    };
+
+    /**
+     * Helper type that wraps an object of type T and records its "in flight" size to
+     * http_plugin_impl::bytes_in_flight using RAII semantics
+     *
+     * @tparam T - the contained Type
+     */
+    template<typename T>
+    struct in_flight {
+        in_flight(T&& object, http_plugin_state& plugin_state)
+        :_object(std::move(object))
+        ,_plugin_state(plugin_state)
+        {
+            _count = detail::in_flight_sizeof(_object);
+            _plugin_state.bytes_in_flight += _count;
+        }
+
+        ~in_flight() {
+            if (_count) {
+                _plugin_state.bytes_in_flight -= _count;
+            }
+        }
+
+        // No copy constructor, but allow move
+        in_flight(const in_flight&) = delete;
+        in_flight(in_flight&& from)
+        :_object(std::move(from._object))
+        ,_count(from._count)
+        ,_plugin_state(std::move(from._plugin_state))
+        {
+            from._count = 0;
+        }
+
+        // No copy assignment, but allow move
+        in_flight& operator=(const in_flight&) = delete;
+        in_flight& operator=(in_flight&& from) {
+            _object = std::move(from._object);
+            _count = from._count;
+            _plugin_state = from._plugin_state;
```

Comment:
This is making a copy of plugin_state.
You actually don't need the move constructor or the move `operator=`, so you can just declare them both as `= delete`.